cli/config/config.go

```diff
@@ -93,3 +101,168 @@ func (p PxfServiceGroup) validate() error {
+		if err != nil {
+			return nil, err
+		}
+		pxfDeployment.Clusters[currentClusterName] = pxfCluster
+	}
+
+	if err := pxfDeployment.Validate(); err != nil {
+		return nil, err
+	}
+
+	return pxfDeployment.Clusters[clusterName], nil
+}
+
+func readFile(path string) ([]string, error) {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return strings.Split(string(data), "\n"), nil
+}
+
+func parseFile(fileContents []string) (map[string]string, error) {
+	parsedFileKeyVals := make(map[string]string)
+
+	for lineNum, line := range fileContents {
+		line = strings.ToLower(line)
+		line = strings.TrimSpace(line)
+		if line == "" || line[0] == '#' { // Remove any line that is a TOML comment
+			continue
+		}
+		lineKeyVal := strings.Split(line, "=")
+		if len(lineKeyVal) != 2 { // lineKeyVal[0] = key, lineKeyVal[1] = value, anything else is invalid
+			return nil, fmt.Errorf("Line must have only one '=' on line %d: %s\n", lineNum+1, line)
+		}
+		key := strings.TrimSpace(lineKeyVal[0])
+		if key == "" {
+			return nil, fmt.Errorf("Cannot have empty key on line %d: %s\n", lineNum+1, line)
+		}
+		if _, exist := parsedFileKeyVals[key]; exist {
+			return nil, fmt.Errorf("Duplicate key on line %d: %s\n", lineNum+1, key)
+		}
+		parsedFileKeyVals[key] = strings.TrimSpace(lineKeyVal[1])
+	}
+
+	return parsedFileKeyVals, nil
+}
+
+func parseCluster(clusterName string, parsedKeyVal map[string]string) (*PxfCluster, error) {
+	cluster := PxfCluster{Groups: make(map[string]*PxfServiceGroup)}
+
+	for key, val := range parsedKeyVal {
+		tokens := strings.Split(key, ".")
+		switch tokens[0] {
+		case "cluster":
+			if err := fillClusterValue(&cluster, tokens, key, val, clusterName); err != nil {
+				return nil, err
+			}
+		case "group":
+			if err := fillServiceGroupValue(cluster.Groups, tokens, key, val); err != nil {
+				return nil, err
+			}
+		default:
+			gplog.Warn("Ignoring %s because it is not a valid key\n", key)
+		}
+	}
+
+	return &cluster, nil
+}
+
+func fillClusterValue(cluster *PxfCluster, tokens []string, key, val, clusterName string) error {
+	if len(tokens) != 2 {
+		return fmt.Errorf("invalid key: %s\n", key)
+	}
+	switch tokens[1] {
+	case "name":
+		err := assertNotEmpty(val, "cluster name")
+		if err != nil {
+			return err
+		}
+		if val != clusterName {
+			return fmt.Errorf("cluster name does not equal directory name of cluster")
+		}
+		cluster.Name = val
+	case "collocated":
+		err := assertNotEmpty(val, "cluster collocated")
+		if err != nil {
+			return err
+		}
+		boolVal, err := strconv.ParseBool(val)
+		if err != nil {
+			return fmt.Errorf("cluster collocated must be a boolean (true/false): %s\n", val)
+		}
+		cluster.Collocated = boolVal
+	case "endpoint":
+		cluster.Endpoint = val
+	default:
+		gplog.Warn("Ignoring %s because it is not a valid key for cluster\n", key)
+	}
+
+	return nil
+}
+
+func fillServiceGroupValue(groups map[string]*PxfServiceGroup, tokens []string, key, val string) error {
```

Comment:
to make it more "object-oriented"
```suggestion
func setGroupProperty (cluster *PxfCluster) (tokens []string, key, val string) error {
```